File: pkg/needle/needle_test.go
package needle_test

import (
	"bytes"
	"testing"

	"github.com/levis9527/jfs/pkg/needle"
)

func TestEncodeDecode(t *testing.T) {
	data := []byte("hello-jfs-small-file")
	buf, n, err := needle.Encode(12345, 99, data)
	if err != nil {
		t.Fatal(err)
	}
	if n.TotalSize%needle.PaddingSize != 0 {
		t.Fatalf("not aligned: %d", n.TotalSize)
	}
	got, err := needle.Decode(buf)
	if err != nil {
		t.Fatal(err)
	}
	if got.Key != 12345 || got.Cookie != 99 {
		t.Fatalf("key/cookie mismatch: %+v", got)
	}
	if !bytes.Equal(got.Data, data) {
		t.Fatalf("data mismatch: %q", got.Data)
	}
}

func TestOffsetAlign(t *testing.T) {
	if needle.NeedleOffset(16) != 2 {
		t.Fatal("NeedleOffset")
	}
	if needle.BlockOffset(2) != 16 {
		t.Fatal("BlockOffset")
	}
	if needle.Align(21) != 24 {
		t.Fatalf("Align got %d", needle.Align(21))
	}
}

func TestCorruptChecksum(t *testing.T) {
	buf, _, err := needle.Encode(1, 1, []byte("abc"))
	if err != nil {
		t.Fatal(err)
	}
	buf[21] ^= 0xff
	if _, err := needle.Decode(buf); err != needle.ErrChecksum {
		t.Fatalf("want checksum err, got %v", err)
	}
}

File: jfs-store/src/test/java/com/levis9527/jfs/needle/NeedleTest.java
package com.levis9527.jfs.needle;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class NeedleTest {
    @Test
    void encodeDecodeRoundTrip() {
        byte[] data = "hello-jfs-small-file".getBytes();
        byte[] buf = Needle.encode(12345L, 99, data);
        assertEquals(0, buf.length % Needle.PADDING_SIZE);
        Needle got = Needle.decode(buf);
        assertEquals(12345L, got.key);
        assertEquals(99, got.cookie);
        assertArrayEquals(data, got.data);
    }

    @Test
    void offsetAlign() {
        assertEquals(2, Needle.needleOffset(16));
        assertEquals(16, Needle.blockOffset(2));
        assertEquals(24, Needle.align(21));
    }

    @Test
    void corruptChecksumDetected() {
        byte[] buf = Needle.encode(1L, 1, "abc".getBytes());
        buf[21] ^= (byte) 0xff;
        assertThrows(NeedleException.class, () -> Needle.decode(buf));
    }

    @Test
    void emptyPayload() {
        byte[] buf = Needle.encode(7L, 3, new byte[0]);
        Needle n = Needle.decode(buf);
        assertEquals(0, n.size);
        assertArrayEquals(new byte[0], n.data);
    }
}
